src/wasm/wasm.cpp

```diff
@@ -174,85 +172,82 @@ handleUnreachable(Block* block,
   }
 }
 
-void Block::finalize() {
-  if (list.size() == 0) {
-    type = Type::none;
-    return;
-  }
-  // The default type is what is at the end. Next we need to see if breaks and/
-  // or unreachability change that.
-  type = list.back()->type;
-  if (!name.is()) {
-    // Nothing branches here, so this is easy.
-    handleUnreachable(this, NoBreak);
-    return;
-  }
+void Block::finalize(std::optional<Type> type_, Breakability breakability) {
+  if (type_) {
+    type = *type_;
+    if (type == Type::none && list.size() > 0) {
+      handleUnreachable(this, breakability);
+    }
 
-  // The default type is according to the value that flows out.
-  BranchUtils::BranchSeeker seeker(this->name);
-  Expression* temp = this;
-  seeker.walk(temp);
-  if (seeker.found) {
-    // Calculate the supertype of the branch types and the flowed-out type. If
-    // there is no supertype among the available types, assume the current type
-    // is already correct. TODO: calculate proper LUBs to compute a new correct
-    // type in this situation.
-    seeker.types.insert(type);
-    type = Type::getLeastUpperBound(seeker.types);
   } else {
-    // There are no branches, so this block may be unreachable.
-    handleUnreachable(this, NoBreak);
-  }
-}
-
-void Block::finalize(Type type_) {
-  type = type_;
-  if (type == Type::none && list.size() > 0) {
-    handleUnreachable(this);
-  }
-}
+    if (list.size() == 0) {
+      type = Type::none;
+      return;
+    }
+    // The default type is what is at the end. Next we need to see if breaks
+    // and/ or unreachability change that.
+    type = list.back()->type;
+    if (!name.is()) {
+      // Nothing branches here, so this is easy.
+      handleUnreachable(this, NoBreak);
+      return;
+    }
 
-void Block::finalize(Type type_, Breakability breakability) {
-  type = type_;
-  if (type == Type::none && list.size() > 0) {
-    handleUnreachable(this, breakability);
+    // The default type is according to the value that flows out.
+    BranchUtils::BranchSeeker seeker(this->name);
+    Expression* temp = this;
+    seeker.walk(temp);
+    if (seeker.found) {
+      // Calculate the supertype of the branch types and the flowed-out type. If
+      // there is no supertype among the available types, assume the current
+      // type is already correct. TODO: calculate proper LUBs to compute a new
+      // correct type in this situation.
```

Comment:
This comment looks extremely stale since we do calculate LUBs now.

Reply:
Fixed.